### PR TITLE
Logger pipeline

### PR DESCRIPTION
**Pull request by heyams**
// TODO need to update azure doc to use logBodies instead of logName in Telemetry Processor sample

### CODE REVIEW

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/ExporterUtil.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.AttributeType;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import java.util.List;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+
+// common string constants and static methods shared among exporters
+final class ExporterUtil {
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/ExporterUtil.java:38]: there's very little here related to logging, i'd suggest keeping it in span exporter, and duplicating the small amount of code needed in the logging exporter

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/LoggerExporterCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  private static List<Configuration.ProcessorConfig> reverseProcessorConfigs(
+      Configuration configuration) {
+    List<Configuration.ProcessorConfig> processors =
+        configuration.preview.processors.stream()
+            .filter(
+                processor ->
+                    processor.type != Configuration.ProcessorType.METRIC_FILTER
+                        && processor.type != Configuration.ProcessorType.SPAN)
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/LoggerExporterCustomizer.java:92]: does this work?
```suggestion
                    processor.type == Configuration.ProcessorType.LOG
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/LoggerExporterCustomizer.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      }
+    }
+
+    // operation name need to be captured while on the main thread
+    // before passing off to the BatchLogProcessor
+    return new AiOperationNameLogWrappingProcessor(
+        BatchLogProcessor.builder(logExporter).setMaxExportBatchSize(1).build());
+  }
+
+  private static List<Configuration.ProcessorConfig> reverseProcessorConfigs(
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/LoggerExporterCustomizer.java:85]: and I think the purpose of reversing is more clearer if moved inside of `createLogExporter`

```suggestion
  private static List<Configuration.ProcessorConfig> getLogProcessorConfigs(
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/OpenTelemetryConfigurer.java [hunk 4]
```diff
@@ -160,13 +151,26 @@ private static BatchSpanProcessor createExporter(Configuration configuration) {
...
+    List<ProcessorConfig> processors =
+        configuration.preview.processors.stream()
+            .filter(
+                processor ->
+                    processor.type != Configuration.ProcessorType.METRIC_FILTER
+                        && processor.type != Configuration.ProcessorType.LOG)
+            .collect(Collectors.toCollection(ArrayList::new));
+    // Reversing the order of processors before passing it to Span/Log processor
+    Collections.reverse(processors);
+    return processors;
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/OpenTelemetryConfigurer.java:173]: similar comments here as above for logging

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java [hunk 5]
```diff
@@ -747,12 +748,12 @@ public void validate(ProcessorType processorType, IncludeExclude includeExclude)
     }
 
     private void validAttributeProcessorIncludeExclude(IncludeExclude includeExclude) {
-      if (attributes.isEmpty() && spanNames.isEmpty()) {
+      if (attributes.isEmpty() && spanNames.isEmpty() && logBodies.isEmpty()) {
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java:751]: can we pass into the method whether it's a span or log processor, so we can validate and log more precisely?

File: test/smoke/testApps/TraceLogBackUsingAgent/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLogBackTest.java [hunk 6]
```diff
@@ -58,7 +58,6 @@ public void testTraceLogBack() throws Exception {
     assertEquals("This is logback warn.", md1.getMessage());
     assertEquals(SeverityLevel.Warning, md1.getSeverityLevel());
     assertEquals("Logger", md1.getProperties().get("SourceType"));
-    assertEquals("WARN", md1.getProperties().get("LoggingLevel"));
```
Comment (trask) [inline test/smoke/testApps/TraceLogBackUsingAgent/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLogBackTest.java:61]: can you find out what .NET does, if they add some kind of `level` property to customDimensions, or if their users rely only on "severity level" kusto column?

File: test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java [hunk 7]
```diff
@@ -44,8 +44,10 @@ public void doMostBasicTest() throws Exception {
     assertTrue(telemetry.rd.getSuccess());
     // Log processor test
     List<MessageData> logs = mockedIngestion.getMessageDataInRequest();
-    MessageData md1 = logs.get(0);
-    assertEquals("smoketestappcontroller::INFO", md1.getMessage());
+    if (!logs.isEmpty()) {
+      MessageData md1 = logs.get(0);
+      assertEquals("This is log from SpringBootApp.", md1.getMessage());
+    }
```
Comment (trask) [inline test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:50]: do u know what caused this change in behavior?

File: agent/agent-tooling/src/test/resources/ApplicationInsights_SpanProcessor.json [hunk 8]
```diff
@@ -85,6 +85,22 @@
...
       },
+      {
+        "type": "log",
+        "include": {
+          "matchType": "regexp",
+          "logBodies": [
+            ".*password.*"
+          ]
+        },
+        "name": {
```
Comment (trask) [inline agent/agent-tooling/src/test/resources/ApplicationInsights_SpanProcessor.json:96]: `name` -> `body`, and should be populating log body field

File: agent/agent-tooling/src/test/resources/ApplicationInsights_SpanProcessor.json [hunk 9]
```diff
@@ -85,6 +85,22 @@
...
+          "fromAttributes": [
+            "loggerName"
```
Comment (trask) [inline agent/agent-tooling/src/test/resources/ApplicationInsights_SpanProcessor.json:98]: this doesn't look like the otel attribute name, so I'm surprised this is working?

**Review by trask — COMMENTED** [review]
quick initial review

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/LoggerExporterCustomizer.java (hunk 2 above)
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/LoggerExporterCustomizer.java:92]: no, Log supports both ProcessorType.LOG and ProcessorType.Attribute. Similar to reverse method in span.

File: test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java (hunk 7 above)
Comment (heyams) [reply test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:50]: this is message body. I think it's upstream that changed it?

File: agent/agent-tooling/src/test/resources/ApplicationInsights_SpanProcessor.json (hunk 9 above)
Comment (heyams) [reply agent/agent-tooling/src/test/resources/ApplicationInsights_SpanProcessor.json:98]: loggerName is different from logName.. i changed logName to logBodies for matching a few lines above..

Comment (heyams) [reply agent/agent-tooling/src/test/resources/ApplicationInsights_SpanProcessor.json:98]: let's chat about this and the one above.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java (hunk 5 above)
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java:751]: It's type Attribute which is applicable to both Log and Span. We don't know yet if it's log or span if you check the call stack.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/LoggerExporterCustomizer.java (hunk 2 above)
Comment (trask) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/LoggerExporterCustomizer.java:91]: ah, then let's change to check those two types, instead of checking `not` the other two types

File: test/smoke/testApps/TraceLogBackUsingAgent/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLogBackTest.java (hunk 6 above)
Comment (heyams) [reply test/smoke/testApps/TraceLogBackUsingAgent/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLogBackTest.java:61]: Confirmed with Cijo, they don't send logginglevel.. only severityLevel.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java (hunk 5 above)
Comment (trask) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java:751]: oh ya, got it 👍

File: test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java (hunk 7 above)
Comment (trask) [reply test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:50]: oh, this is `TelemetryProcessor` test, it's supposed to update the log boddy with `smoketestappcontroller::INFO`, see the telemetry json config for this test

File: agent/agent-tooling/src/test/java/com/microsoft/applicationinsights/agent/internal/processors/MockLogExporter.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+import com.microsoft.applicationinsights.agent.internal.exporter.LoggerExporter;
+import com.microsoft.applicationinsights.agent.internal.telemetry.TelemetryClient;
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.logs.data.LogData;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class MockLogExporter extends LoggerExporter {
```
Comment (trask) [inline agent/agent-tooling/src/test/java/com/microsoft/applicationinsights/agent/internal/processors/MockLogExporter.java:32]: `MockLogExporter` -> `MockLoggerExporter` (just to match `LoggerExporter`)

File: agent/agent-tooling/src/test/java/com/microsoft/applicationinsights/agent/internal/processors/MockLogData.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.sdk.common.InstrumentationLibraryInfo;
+import io.opentelemetry.sdk.logs.data.Body;
+import io.opentelemetry.sdk.logs.data.LogData;
+import io.opentelemetry.sdk.logs.data.Severity;
+import io.opentelemetry.sdk.resources.Resource;
+import javax.annotation.Nullable;
+
+public class MockLogData implements LogData {
```
Comment (trask) [inline agent/agent-tooling/src/test/java/com/microsoft/applicationinsights/agent/internal/processors/MockLogData.java:33]: let's rename to `TestLogData`, hopefully otel will provide the same at some point similar to `TestSpanData` (and it's not really a mock)

File: test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java (hunk 7 above)
Comment (heyams) [reply test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:50]: good catch, i modified the json.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/Exporter.java [hunk 12]
```diff
@@ -1014,7 +917,7 @@ private static void addLinks(AbstractTelemetryBuilder telemetryBuilder, List<Lin
   }
 
   private static void setExtraAttributes(
-      AbstractTelemetryBuilder telemetryBuilder, Attributes attributes) {
+      AbstractTelemetryBuilder telemetryBuilder, Attributes attributes, Logger logger) {
```
[the remark this answers is not in the document]
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/Exporter.java:920]: don't understand this comment.  it's used in line 994: ` logger.warn("unexpected attribute type: {}", type);`

File: test/smoke/testApps/TraceLog4j2UsingAgent/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLog4j2Test.java [hunk 13]
```diff
@@ -60,25 +60,19 @@ public void testTraceLog4j2() throws Exception {
...
-    assertEquals("Logger", md2.getProperties().get("SourceType"));
-    assertEquals("ERROR", md2.getProperties().get("LoggingLevel"));
     assertEquals("smoketestapp", md1.getProperties().get("LoggerName"));
     assertNotNull(md1.getProperties().get("ThreadName"));
     assertEquals(4, md2.getProperties().size());
 
     assertEquals("This is log4j2 fatal.", md3.getMessage());
     assertEquals(SeverityLevel.Critical, md3.getSeverityLevel());
-    assertEquals("Logger", md3.getProperties().get("SourceType"));
-    assertEquals("FATAL", md3.getProperties().get("LoggingLevel"));
```
Comment (trask) [inline test/smoke/testApps/TraceLog4j2UsingAgent/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLog4j2Test.java:81]: can we bring these verifications back now?

File: test/smoke/testApps/InheritedAttributes/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java [hunk 14]
```diff
@@ -60,6 +60,6 @@ public void test() throws Exception {
     assertEquals("smoketestapp", md.getProperties().get("LoggerName"));
     assertNotNull(md.getProperties().get("ThreadName"));
     assertEquals("z", rd.getProperties().get("tenant"));
-    assertEquals(5, md.getProperties().size());
+    assertEquals(4, md.getProperties().size());
```
Comment (trask) [inline test/smoke/testApps/InheritedAttributes/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:63]: I don't remember why this change?

File: test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java [hunk 15]
```diff
@@ -44,8 +44,10 @@ public void doMostBasicTest() throws Exception {
     assertTrue(telemetry.rd.getSuccess());
     // Log processor test
     List<MessageData> logs = mockedIngestion.getMessageDataInRequest();
-    MessageData md1 = logs.get(0);
-    assertEquals("smoketestappcontroller::INFO", md1.getMessage());
+    if (!logs.isEmpty()) {
+      MessageData md1 = logs.get(0);
+      assertEquals("testValue1::testValue2", md1.getMessage());
+    }
```
Comment (trask) [inline test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:50]: do we need the conditional?

```suggestion
    MessageData md1 = logs.get(0);
    assertEquals("testValue1::testValue2", md1.getMessage());
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AgentProcessor.java [hunk 16]
```diff
@@ -125,16 +121,17 @@ private boolean checkAttributes(SpanData span) {
 
   public static class RegexpIncludeExclude extends IncludeExclude {
 
-    private final List<Pattern> spanPatterns;
+    private final List<Pattern> patterns;
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AgentProcessor.java:124]:
```suggestion
    // pattern to match against span names or log bodies
    private final List<Pattern> patterns;
```

File: test/smoke/testApps/InheritedAttributes/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java (hunk 14 above)
Comment (heyams) [reply test/smoke/testApps/InheritedAttributes/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:63]: i checked main branch; 

it returned. 

`{"LoggerName":"smoketestapp","LoggingLevel":"INFO","SourceType":"Logger","ThreadName":"http-nio-8080-exec-2","tenant":"z"}`

```
ThreadContext.getRequestTelemetryContext()
        .getHttpRequestTelemetry()
        .getProperties()
        .put("tenant", "z");
```

Comment (trask) [reply test/smoke/testApps/InheritedAttributes/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:63]: why doesn't the same thing happen in this PR branch? just want to make sure we didn't inadvertently break something

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+    setExtraAttributes(telemetryBuilder, attributes);
+
+    telemetryBuilder.setSeverityLevel(toSeverityLevel(log.getSeverity()));
+    telemetryBuilder.setMessage(log.getBody().asString());
+
+    // set message-specific properties
+    setLoggerProperties(
+        telemetryBuilder,
+        log.getInstrumentationLibraryInfo().getName(),
+        attributes.get(SemanticAttributes.THREAD_NAME),
```
Comment (lzchen) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:139]: Should we hold off on this until [this](https://github.com/open-telemetry/opentelemetry-specification/pull/2389) is merged?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+        failure = true;
+      }
+    }
+
+    return failure ? CompletableResultCode.ofFailure() : CompletableResultCode.ofSuccess();
+  }
+
+  @Override
+  public CompletableResultCode flush() {
+    return CompletableResultCode.ofSuccess();
```
Comment (lzchen) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:96]: Is this TODO?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java [hunk 19]
```diff
@@ -0,0 +1,359 @@
...
+
+  private void trackMessage(LogData log) {
+    MessageTelemetryBuilder telemetryBuilder = telemetryClient.newMessageTelemetryBuilder();
+
+    Attributes attributes = log.getAttributes();
+
+    // set standard properties
+    telemetryBuilder.setTime(FormattedTime.offSetDateTimeFromEpochNanos(log.getEpochNanos()));
+    setOperationTags(telemetryBuilder, log);
+    setSampleRate(telemetryBuilder, log);
```
Comment (lzchen) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:129]: Is this defined anywhere in common schema?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java [hunk 20]
```diff
@@ -0,0 +1,359 @@
...
+
+    telemetryBuilder.setSeverityLevel(toSeverityLevel(log.getSeverity()));
+    telemetryBuilder.setMessage(log.getBody().asString());
+
+    // set message-specific properties
+    setLoggerProperties(
+        telemetryBuilder,
+        log.getInstrumentationLibraryInfo().getName(),
+        attributes.get(SemanticAttributes.THREAD_NAME),
+        log.getSeverity());
```
Comment (lzchen) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:140]: What's the difference between `telemetryBuilder.addProperty("LoggingLevel", mapSeverityToLoggingLevel(severity).toString());` and `telemetryBuilder.setSeverityLevel(toSeverityLevel(log.getSeverity()));`?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java [hunk 21]
```diff
@@ -0,0 +1,359 @@
...
+
+  private void trackMessageAsException(LogData log, String stack) {
+    ExceptionTelemetryBuilder telemetryBuilder = telemetryClient.newExceptionTelemetryBuilder();
+    Attributes attributes = log.getAttributes();
+
+    // set standard properties
+    setOperationTags(telemetryBuilder, log);
+    telemetryBuilder.setTime(FormattedTime.offSetDateTimeFromEpochNanos(log.getEpochNanos()));
+    setSampleRate(telemetryBuilder, log);
+    setExtraAttributes(telemetryBuilder, attributes);
```
Comment (lzchen) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:154]: Isn't the stack coming in as an attribute as well? Would this set it twice?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java [hunk 22]
```diff
@@ -0,0 +1,359 @@
...
+
+    telemetryBuilder.setExceptions(Exceptions.minimalParse(stack));
```
Comment (lzchen) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:156]: Isn't the stack coming in as an attribute as well? Would this set it twice?

Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:156]: stack is part of the attribute of the log and we check if such an attribute exists, then track it as exception, not trace.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java (hunk 20 above)
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:140]: they're two different properties. LoggingLevel is something we sent in the cusomDimensions and severity is a property required by the common schema.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java (hunk 19 above)
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:129]: not defined yet currently.. but it should be defined though since we send this in all telemetry..

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java (hunk 18 above)
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:96]: yes, need to implement it in standalone exporter later.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java (hunk 17 above)
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:139]: we have been sending thread_name to trace table.

File: test/smoke/testApps/InheritedAttributes/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java [hunk 23]
```diff
@@ -59,7 +59,7 @@ public void test() throws Exception {
     assertEquals("INFO", md.getProperties().get("LoggingLevel"));
     assertEquals("smoketestapp", md.getProperties().get("LoggerName"));
     assertNotNull(md.getProperties().get("ThreadName"));
-    assertEquals("z", rd.getProperties().get("tenant"));
+    assertEquals("z", md.getProperties().get("tenant"));
```
Comment (trask) [inline test/smoke/testApps/InheritedAttributes/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:62]: good catch 👍

File: test/smoke/testApps/TraceLog4j1_2/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLog4j12Test.java [hunk 24]
```diff
@@ -78,7 +78,7 @@ public void testTraceLog4j12() throws Exception {
     assertEquals("This is log4j1.2 fatal.", md3.getMessage());
     assertEquals(SeverityLevel.Critical, md3.getSeverityLevel());
     assertEquals("Logger", md3.getProperties().get("SourceType"));
-    assertEquals("FATAL", md3.getProperties().get("LoggingLevel"));
+    assertEquals("ERROR", md3.getProperties().get("LoggingLevel"));
```
Comment (trask) [inline test/smoke/testApps/TraceLog4j1_2/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLog4j12Test.java:81]: we can change this back now with latest change
```suggestion
    assertEquals("FATAL", md3.getProperties().get("LoggingLevel"));
```

File: test/smoke/testApps/TraceLog4j1_2UsingAgent/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLog4j12Test.java [hunk 25]
```diff
@@ -78,8 +78,8 @@ public void testTraceLog4j12() throws Exception {
     assertEquals("This is log4j1.2 fatal.", md3.getMessage());
     assertEquals(SeverityLevel.Critical, md3.getSeverityLevel());
     assertEquals("Logger", md3.getProperties().get("SourceType"));
-    assertEquals("FATAL", md3.getProperties().get("LoggingLevel"));
-    assertEquals("smoketestapp", md3.getProperties().get("LoggerName"));
+    assertEquals("ERROR", md3.getProperties().get("LoggingLevel"));
+    assertEquals("smoketestapp", md2.getProperties().get("LoggerName"));
```
Comment (trask) [inline test/smoke/testApps/TraceLog4j1_2UsingAgent/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLog4j12Test.java:82]:
```suggestion
    assertEquals("FATAL", md3.getProperties().get("LoggingLevel"));
    assertEquals("smoketestapp", md3.getProperties().get("LoggerName"));
```

File: test/smoke/testApps/TraceLog4j2/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLog4j2Test.java [hunk 26]
```diff
@@ -78,7 +78,7 @@ public void testTraceLog4j2() throws Exception {
     assertEquals("This is log4j2 fatal.", md3.getMessage());
     assertEquals(SeverityLevel.Critical, md3.getSeverityLevel());
     assertEquals("Logger", md3.getProperties().get("SourceType"));
-    assertEquals("FATAL", md3.getProperties().get("LoggingLevel"));
+    assertEquals("ERROR", md3.getProperties().get("LoggingLevel"));
```
Comment (trask) [inline test/smoke/testApps/TraceLog4j2/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/TraceLog4j2Test.java:81]:
```suggestion
    assertEquals("FATAL", md3.getProperties().get("LoggingLevel"));
```

**Review by trask — APPROVED** [review]
🎉

let's revisit the java.util.logging level change prior to release

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java [hunk 27]
```diff
@@ -223,7 +223,7 @@ static void setExtraAttributes(AbstractTelemetryBuilder telemetryBuilder, Attrib
           if (stringKey.startsWith("thread.")) {
             return;
           }
-          if (stringKey.equals("exception.")) {
+          if (stringKey.startsWith("exception.")) {
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java:226]: 👍

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/InheritedAttributesLogProcessor.java [hunk 28]
```diff
@@ -77,6 +79,9 @@ public void emit(LogData log) {
     AttributesBuilder builder = log.getAttributes().toBuilder();
     for (AttributeKey<?> inheritedAttributeKey : inheritedAttributes) {
       Object value = readableSpan.getAttribute(inheritedAttributeKey);
+      if (inheritedAttributeKey == AI_OPERATION_NAME_KEY) {
+        value = AiOperationNameSpanProcessor.getOperationName(readableSpan);
+      }
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/InheritedAttributesLogProcessor.java:83]: 👍

```suggestion
      if (value == null && inheritedAttributeKey == AI_OPERATION_NAME_KEY) {
        value = AiOperationNameSpanProcessor.getOperationName(readableSpan);
      }
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/InheritedAttributesLogProcessor.java [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+      return;
+    }
+
+    ReadableSpan readableSpan = (ReadableSpan) currentSpan;
+    AttributesBuilder builder = log.getAttributes().toBuilder();
+    for (AttributeKey<?> inheritedAttributeKey : inheritedAttributes) {
+      Object value = readableSpan.getAttribute(inheritedAttributeKey);
+      if (value == null && inheritedAttributeKey == AI_OPERATION_NAME_KEY) {
+        value = AiOperationNameSpanProcessor.getOperationName(readableSpan);
+      }
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/InheritedAttributesLogProcessor.java:83]:
```suggestion
      Object value;
      if (inheritedAttributeKey == AI_OPERATION_NAME_KEY) {
        value = AiOperationNameSpanProcessor.getOperationName(readableSpan);
      } else {
        value = readableSpan.getAttribute(inheritedAttributeKey);
      }
```
